lib/pdk/config/json_schema_namespace.rb

```diff
@@ -0,0 +1,123 @@
+      # TODO: What if the document is not valid?!?!? We can never save then
+
+      # The JSON Schema for the namespace
+      #
+      # @return [Hash]
+      def schema
+        document_schema.schema
+      end
+
+      # Whether the schema is valid but empty.
+      #
+      # @return [Boolean]
+      def empty_schema?
+        document_schema.schema.empty?
+      end
+
+      # Name of all the top level properties for the schema
+      #
+      # @return [String[]]
+      def schema_property_names
+        return [] if schema['properties'].nil?
+        schema['properties'].keys
+      end
+
+      # Extends the to_h namespace method to include unmanaged settings
+      #
+      # @see PDK::Config::Namespace.to_h
+      def to_h
+        # This may seem counter-intuitive but we need to call super first as the settings
+        # may not have been loaded yet, which means @unmanaged_settings will be empty.
+        # We call super first to force any file loading and then merge the unmanaged settings
+        settings_hash = super
+        @unmanaged_settings = {} if @unmanaged_settings.nil?
+        @unmanaged_settings.merge(settings_hash)
+      end
+
+      # Validates a document (Hash table) against the schema
+      #
+      # @return [Boolean]
+      def validate_document!(document)
+        ::JSON::Validator.validate!(schema, document)
+      end
+
+      # Sets the list of unmanaged settings. For subclass use only
+      #
+      # @param value [Hash<String, Object]] A hashtable of all unmanaged settings which will be persisted, but not visible
+      #
+      # @private
+      def assign_unmanaged_settings(value)
+        @unmanaged_settings = value
+      end
+
+      private
+
+      # Override the create_setting method to always fail. This is called
+      # to dyanmically add settings. However as we're using a schema, no
+      # new settings can be created
+      #
+      # @see PDK::Config::Namespace.create_missing_setting
+      #
+      # @private
+      def create_missing_setting(key, _initial_value = nil)
+        raise ArgumentError, _("Setting '#{key}' does not exist'")
+      end
+
+      # Create a valid, but empty schema
+      #
+      # @return [JSON::Schema]
+      def create_empty_schema
+        require 'json-schema'
+        ::JSON::Schema.new({}, 'http://json-schema.org/draft-06/schema#')
+      end
+
+      # Lazily retrieve the JSON schema from disk for this namespace
+      #
+      # @return [JSON::Schema]
+      def document_schema
+        return @document_schema unless @document_schema.nil?
+
+        require 'json-schema'
+
+        if @schema_file.nil?
+          @document_schema = create_empty_schema
+          return @document_schema
+        end
+        unless PDK::Util::Filesystem.file?(@schema_file)
+          @document_schema = create_empty_schema
+          raise PDK::Config::LoadError, _('Unable to open %{file} for reading. File does not exist') % {
+            file: @schema_file,
+          }
```

Comment:
Does this still need to `create_empty_schema` if it's going to `raise` right after? Is there a use case where the exception is caught and `@document_schema` is expected to be set?

Reply:
(I see it does the same thing down below so I assume that side effect is intended?)

Reply:
It is possible for namespace to not care that a schema failed. So I needed some kind of value.  That said, I could perhaps assign it better/earlier? so it doesn't look as strange.

Reply:
This case could potentially just be another `rescue` for whatever `JSON::Schema::Reader` raises if the file doesn't exist? (Unless of course it just raises `::JSON::Schema::JsonParseError` :/) That might clean up the rest of the flow control here.

Reply:
Done.

Reply:
Fixed.  Uses an empty schema as the "default" value.